build-logic/src/main/java/org/gradle/android/plugin/AndroidVersionsPlugin.java

```diff
@@ -0,0 +1,62 @@
+package org.gradle.android.plugin;
+
+import org.gradle.api.Plugin;
+import org.gradle.api.Project;
+import org.gradle.api.model.ObjectFactory;
+import org.gradle.api.plugins.JavaPluginExtension;
+import org.gradle.api.tasks.SourceSet;
+import org.gradle.api.tasks.TaskProvider;
+import org.gradle.api.tasks.testing.Test;
+
+import java.util.ArrayList;
+
+public class AndroidVersionsPlugin implements Plugin<Project> {
+
+    @Override
+    public void apply(Project project) {
+        ObjectFactory objects = project.getObjects();
+        var androidVersionsContainer = objects.domainObjectContainer(AndroidVersionDomainObject.class,
+            name -> objects.newInstance(AndroidVersionDomainObject.class, name));
+        project.getExtensions().add("androidVersions", androidVersionsContainer);
+        var generateTask = project.getTasks().register("generateVersions", GenerateVersionsTask.class);
+        generateTask.configure(t -> {
+            t.getSupportedVersions().set(androidVersionsContainer);
+            t.getVersion().set((String) project.getVersion());
+        });
+
+        var sourceSets = project.getExtensions().getByType(JavaPluginExtension.class).getSourceSets();
```

Comment:
This plugin expects `JavaPluginExtension` to exist in order to work properly.  It should either apply the java plugin, or it should respond to the java plugin being applied (i.e. `pluginManager.withPlugin()`).  Otherwise, you have an ordering problem (i.e. java plugin _must_ be applied before this plugin in the applying project).

Reply:
Agreed.